mlflow/tracking/fluent.py

```diff
@@ -3053,32 +3053,67 @@ def setup_autologging(module):
             register_post_import_hook(setup_autologging, "pyspark.ml", overwrite=True)
 
 
-_ACTIVE_MODEL_ID = ContextVar("active_model_id", default=None)
+class ActiveModelContext:
+    """
+    The context of the active model.
+
+    Args:
+        model_id: The ID of the active model.
+        set_by_user: Whether the active model was set by the user or not.
+    """
+
+    def __init__(self, model_id: Optional[str] = None, set_by_user: bool = False):
+        self.model_id = model_id
+        self.set_by_user = set_by_user
+
+    @property
+    def model_id(self) -> Optional[str]:
+        return self._model_id or MLFLOW_ACTIVE_MODEL_ID.get()
+
+    @model_id.setter
+    def model_id(self, value: Optional[str]):
+        self._model_id = value
+
+    @property
+    def set_by_user(self) -> bool:
+        return self._set_by_user
+
+    @set_by_user.setter
+    def set_by_user(self, value: bool):
+        self._set_by_user = value
+
+
+_ACTIVE_MODEL_CONTEXT = ContextVar(
+    "active_model_context",
+    default=ActiveModelContext(),
+)
 
 
 class ActiveModel(LoggedModel):
     """
     Wrapper around :py:class:`mlflow.entities.LoggedModel` to enable using Python ``with`` syntax.
     """
 
-    def __init__(self, logged_model: LoggedModel):
+    def __init__(self, logged_model: LoggedModel, set_by_user: bool):
         super().__init__(**logged_model.to_dictionary())
-        self.last_active_model_id = _ACTIVE_MODEL_ID.get()
-        _ACTIVE_MODEL_ID.set(self.model_id)
-        MLFLOW_ACTIVE_MODEL_ID.set(self.model_id)
+        self.last_active_model_context = _ACTIVE_MODEL_CONTEXT.get()
+        self.last_active_model_id_env_var = MLFLOW_ACTIVE_MODEL_ID.get()
+        _ACTIVE_MODEL_CONTEXT.set(ActiveModelContext(self.model_id, set_by_user))
+        _update_active_model_id_env_var(self.model_id)
```

Comment:
We still can't get rid of explicit set for MLFLOW_ACTIVE_MODEL_ID, since it could be different from ActiveModelContext._model_id.